google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/CloudSdkPanel.java

```diff
@@ -70,46 +71,82 @@ public CloudSdkPanel() {
         .addDocumentListener(new DocumentAdapter() {
           @Override
           protected void textChanged(DocumentEvent event) {
-            checkSdk();
+            checkSdkInBackground();
           }
         });
 
-    checkSdk();
+    checkSdkInBackground();
   }
 
-  private void checkSdk() {
-    String path = cloudSdkDirectoryField.getText();
+  private void checkSdkInBackground() {
+    final String path = cloudSdkDirectoryField.getText();
+    ApplicationManager.getApplication().executeOnPooledThread(new CloudSdkCheckerRunnable(path));
+  }
 
+  @VisibleForTesting
+  protected void checkSdk(String path) {
     if (StringUtil.isEmpty(path)) {
-      warningIcon.setVisible(true);
-      warningMessage.setVisible(true);
-      warningMessage.setText(
-          createErrorMessageWithLink(
-              GctBundle.message("appengine.cloudsdk.location.missing.message")));
-
+      String warningText = appendCloudSdkDownloadMessage(
+              GctBundle.message("appengine.cloudsdk.location.missing.message"));
+      showWarning(warningText, false /* setSdkDirectoryErrorState */);
       return;
     }
 
-    try {
-      new CloudSdk.Builder()
-          .sdkPath(Paths.get(path))
-          .build()
-          .validateCloudSdk();
-
-      cloudSdkDirectoryField.getTextField().setForeground(JBColor.black);
-      warningIcon.setVisible(false);
-      warningMessage.setVisible(false);
-    } catch (AppEngineException aee) {
-      cloudSdkDirectoryField.getTextField().setForeground(JBColor.red);
-      warningIcon.setVisible(true);
-      warningMessage.setVisible(true);
-      warningMessage.setText(
-          createErrorMessageWithLink(
-              GctBundle.message("appengine.cloudsdk.location.invalid.message")));
+    Set<CloudSdkValidationResult> validationResults =
+        CloudSdkService.getInstance().validateCloudSdk(Paths.get(path));
+
+    // if there are any validation errors, show the first one to the user
```

Comment:
It could look something like this (though be aware that these two validation messages won't ever actually occur together. What do you think?

![screenshot from 2016-11-16 15 13 44](https://cloud.githubusercontent.com/assets/1039411/20363970/5d1cf3c6-ac0f-11e6-81ce-ccbe41b1cd01.png)